main/lsp/ExtractVariable.cc

```diff
@@ -290,8 +290,17 @@ class LocSearchWalk {
         auto &send = ast::cast_tree_nonnull<ast::Send>(tree);
         if (send.fun == core::Names::orAsgn() || send.fun == core::Names::andAsgn() ||
             send.fun == core::Names::opAsgn()) {
+            ENFORCE(send.numPosArgs() >= 1);
             skipLocExact(send.getPosArg(0).loc());
         }
+
+        if (auto constantLit = ast::cast_tree<ast::ConstantLit>(send.recv)) {
+            if (constantLit->symbol == core::Symbols::Regexp() && send.fun == core::Names::new_()) {
+                ENFORCE(send.numPosArgs() >= 2);
+                skipLocRange(send.getPosArg(0).loc());
+                skipLocRange(send.getPosArg(1).loc());
+            }
+        }
```

Comment:
A couple of questions when looking at this code:

1. Is it possible to call this on user-written calls to `Regexp.new`, and not just desugared calls from `/.../i` or `%r{...}` etc.?
1. If so, since this is potentially user-written code (as opposed to the `orAsgn` and friends case above), I don't think we can blithely `ENFORCE` that we have two positional args, we have to check properly.
1. Is there some technical reason that we can't extract the regexp -- whether a literal or an actual call to `Regexp.new` -- into a variable?
1. Why are we only skipping the locs for the args, and not for the entire send?

Reply:
1. No, we desugar `Regexp.new("abc", "i")` differently from `/abc/i`.
2. I added a test for the single arg passed to an explicit call to `Regexp.new`, to ensure this ENFORCE doesn't fire there
3. I'm trying to prevent extracting of `abc` and `i` on its own in `/abc/i`, since that would lead to `/newVariable/i` or `/abc/newVariable`, which are both invalid. Extracting the entire regex, or the arguments to an explicit call to `Rexexp.new` is fine.
4. It's fine to extract `/abc/i` as a whole.